src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzer.cs

```diff
@@ -0,0 +1,136 @@
+    {
+        private static readonly LocalizableString s_localizableTitle = CreateLocalizableResourceString(nameof(DoNotDeclareCompilerFeatureInAssemblyWithUnsupportedTargetFrameworkRuleTitle));
+        private static readonly LocalizableString s_localizableDescription = CreateLocalizableResourceString(nameof(DoNotDeclareCompilerFeatureInAssemblyWithUnsupportedTargetFrameworkRuleDescription));
+        private const string HelpLinkUri = "https://github.com/dotnet/roslyn-analyzers/blob/main/docs/rules/RS1041.md";
+
+        public static readonly DiagnosticDescriptor DoNotDeclareCompilerFeatureInAssemblyWithUnsupportedTargetFrameworkStrictRule = new(
+            DiagnosticIds.DoNotRegisterCompilerTypesWithBadTargetFrameworkRuleId,
+            s_localizableTitle,
+            CreateLocalizableResourceString(nameof(DoNotDeclareCompilerFeatureInAssemblyWithUnsupportedTargetFrameworkMessage)),
+            DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            helpLinkUri: HelpLinkUri,
+            customTags: WellKnownDiagnosticTagsExtensions.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+            DoNotDeclareCompilerFeatureInAssemblyWithUnsupportedTargetFrameworkStrictRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var typeProvider = WellKnownTypeProvider.GetOrCreate(context.Compilation);
+                var diagnosticAnalyzer = typeProvider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisDiagnosticsDiagnosticAnalyzer);
+                if (diagnosticAnalyzer is null)
+                    return;
+
+                var targetFrameworkAttribute = typeProvider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeVersioningTargetFrameworkAttribute);
+                if (targetFrameworkAttribute is null)
+                    return;
+
+                AttributeData? appliedTargetFrameworkAttribute = context.Compilation.Assembly.GetAttribute(targetFrameworkAttribute);
+                if (appliedTargetFrameworkAttribute is null)
+                    return;
+
+                if (appliedTargetFrameworkAttribute.ConstructorArguments.IsEmpty)
+                {
+                    return;
+                }
+
+                string displayName;
+                switch (appliedTargetFrameworkAttribute.ConstructorArguments[0].Value as string)
+                {
+                    case ".NETStandard,Version=v1.0":
```

Comment:
The compiler already has the ability to pass MSBuild properties through as analyzer config options. By default we pass none. I'm starting to wonder if we should automatically pass a few ones like `<TargetFramework>` here. There are a number of analyzers that are interested in TFM. Looking at it that way doesn't require an type of semantic model access.

Reply:
Is this done by project-system?

Reply:
This repository has the ability to (and does in some cases) add such properties to the list of analyzer-exposed values. I can convert the analyzer to use this if desired.